deeplearning4j-scaleout/spark/dl4j-spark-ml/src/main/scala/org/deeplearning4j/spark/sql/sources/mnist/MnistRelation.scala

```diff
@@ -31,18 +30,33 @@ case class MnistRelation(imagesFile: String, labelsFile: String, numExamples: In
   override def buildScan(requiredColumns: Array[String]): RDD[Row] = {
     val sc = sqlContext.sparkContext
 
-    val baseData = for (i <- 0 until numExamples) yield {
-      (manager.readLabel(), ArrayUtil.flatten(manager.readImage()))
+    val manager = {
+      // CAVEAT: MnistManager supports only local files at this time.
+      val imagesFile = Paths.get(imagesPath.toUri).toFile.getAbsolutePath
+      val labelsFile = Paths.get(labelsPath.toUri).toFile.getAbsolutePath
+      new MnistManager(imagesFile, labelsFile)
     }
+    try {
+      val cols = {
+        val req = requiredColumns.toSet
+        schema.fieldNames flatMap {
+          case "label" if req("label") => Seq(
+            (pt:Int) => manager.readLabel().toDouble)
+          case "features" if req("features") => Seq(
+            (pt:Int) => Vectors.dense(ArrayUtil.flatten(manager.readImage()).map(_.toDouble)))
+          case _ => Seq.empty
+        }
+      }
 
-    val rowBuilders = requiredColumns.map {
-      case "label" => (pt: (Int, Array[Int])) => Seq(pt._1.toDouble)
-      case "features" => (pt: (Int, Array[Int])) => Seq(Vectors.dense(pt._2.map(_.toDouble)))
-    }
+      val numExamples = manager.getLabels.getCount
 
-    sc.makeRDD(baseData.map(pt => {
-      Row.fromSeq(rowBuilders.map(_(pt)).reduceOption(_ ++ _).getOrElse(Seq.empty))
-    }))
+      sc.makeRDD((0 until numExamples).map(pt => {
```

Comment:
Why is `numExample` necessary here? In this code, `numExamples` always seems to represent all data. So I think that mapping simply sufficient in this use case.